managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -1757,32 +1757,26 @@ long getNumIndividualDeletedEntriesToSkip(long numEntries) {
         try {
             InvidualDeletedMessagesHandlingState state = new InvidualDeletedMessagesHandlingState(markDeletePosition);
             individualDeletedMessages.forEach((r) -> {
-                try {
-                    state.endPosition = r.lowerEndpoint();
-                    if (state.startPosition.compareTo(state.endPosition) <= 0) {
-                        Range<PositionImpl> range = Range.openClosed(state.startPosition, state.endPosition);
-                        long entries = ledger.getNumberOfEntries(range);
-                        if (state.totalEntriesToSkip + entries >= numEntries) {
-                            // do not process further
-                            return false;
-                        }
-                        state.totalEntriesToSkip += entries;
-                        state.deletedMessages += ledger.getNumberOfEntries(r);
-                        state.startPosition = r.upperEndpoint();
-                    } else {
-                        if (log.isDebugEnabled()) {
-                            log.debug("[{}] deletePosition {} moved ahead without clearing deleteMsgs {} for cursor {}",
-                                    ledger.getName(), markDeletePosition, r.lowerEndpoint(), name);
-                        }
+                state.endPosition = r.lowerEndpoint();
+                if (state.startPosition.compareTo(state.endPosition) <= 0) {
+                    Range<PositionImpl> range = Range.openClosed(state.startPosition, state.endPosition);
+                    long entries = ledger.getNumberOfEntries(range);
+                    if (state.totalEntriesToSkip + entries >= numEntries) {
+                        // do not process further
+                        return false;
                     }
-                    return true;
-                } finally {
-                    if (r.lowerEndpoint() instanceof PositionImplRecyclable) {
-                        ((PositionImplRecyclable) r.lowerEndpoint()).recycle();
```

Comment:
We can just recycle r.lowerEndpoint() here.

Reply:
Good suggestion

Reply:
Good suggestion